crates/auth/src/providers/auth0/mod.rs

```diff
@@ -0,0 +1,99 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: AGPL-3.0-only
+
+pub mod routes;
+
+use super::super::session::Session;
+
+use std::{fmt, io};
+
+use axum::http::header::{AUTHORIZATION, USER_AGENT};
+use axum::http::status::InvalidStatusCode;
+use axum::http::StatusCode;
+use oauth2::basic::BasicClient;
+use oauth2::{AuthUrl, ClientId, RedirectUrl, TokenUrl};
+use serde::Deserialize;
+
+pub const LOGIN_URI: &str = "/auth/auth0";
+pub const LOGOUT_URI: &str = "/auth/auth0/logout";
+// pub const AUTHORIZED_URI: &str = "/auth/auth0/authorized";
+pub const AUTHORIZED_URI: &str = "/authorized";
+
+#[derive(Debug)]
+pub(crate) enum ValidateError {
+    Http(ureq::Error),
+    Json(io::Error),
+    InvalidStatusCode(InvalidStatusCode),
+    Auth0(String),
+}
+
+impl fmt::Display for ValidateError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "session {}",
+            match self {
+                ValidateError::Http(e) => format!("HTTP request error: {}", e),
+                ValidateError::Json(e) => format!("JSON deserialization error: {}", e),
+                ValidateError::InvalidStatusCode(e) => format!("invalid status code: {}", e),
+                ValidateError::Auth0(e) => format!("auth0 error: {}", e),
+            }
+        )
+    }
+}
+
+impl std::error::Error for ValidateError {}
+
+#[derive(Deserialize)]
+pub(crate) struct Auth0User {
+    #[serde(rename = "login")]
+    pub username: String,
+    pub id: u64,
+}
+
+//////////////////////// TODO
+pub(crate) async fn validate(session: &Session) -> Result<Auth0User, ValidateError> {
+    #[derive(Deserialize)]
+    struct Error {
+        message: String,
+    }
+
+    let res = ureq::get("https://api.github.com/user")
```

Comment:
This should be doing a `UserInfo` request using the configured OIDC client https://github.com/profianinc/drawbridge/blob/bb43cb593f04030b376143bfe55476f06da025fc/crates/app/src/auth/oidc.rs#L94-L113